### PR TITLE
Enable GCP Flow Logs

### LINKED ISSUE #313: Ability to configure GCP VPC Flow Logs
We need the ability to configure the VPC flow logs on Google Cloud Platform via Gardener to meet security requirements.

Manual configuration of the VPC flow logs is not possible, as it gets overwritten by Gardener on reconciliation.

### PR DESCRIPTION
**Pull request by DockToFuture**
**What this PR does / why we need it**:
This PR enables the usage of vpc flow logs on gcp and gives the possiblity to configure them via the ```InfrastructureConfig```.
**Which issue(s) this PR fixes**:
Fixes #313

**Special notes for your reviewer**:
Requires v="3.4.0" of google terraform provider , see https://github.com/gardener/terraformer/commit/e00e30dda9362786dcc11a89b564f94ede81f5bc.
Needs to be updated in https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-gcp/charts/images.yaml#L5

**Release note**:

```improvement operator
It is now possible to enable VPC Flow Logs on GCP by specifying `spec.provider.infrastructureConfig.networks.flowLogs` section in the `shoot.yaml` file. If the section is not provided then the VPC Flow Logs are disabled.
```

### CODE REVIEW

File: controllers/provider-gcp/charts/internal/gcp-infra/values.yaml [hunk 1]
```diff
@@ -20,6 +20,11 @@ networks:
   pods: 100.96.0.0/11
   workers: 10.250.0.0/19
 # internal: 10.250.112.0/22
+  enableVPCFlowLogs: false
+  flowLog:
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/values.yaml]: Can you comment the whole block to make obvious that this is completely optional and nothing happens by default?

```suggestion
# flowLogs:
```

File: controllers/provider-gcp/charts/internal/gcp-infra/values.yaml [hunk 2]
```diff
@@ -20,6 +20,11 @@ networks:
   pods: 100.96.0.0/11
   workers: 10.250.0.0/19
 # internal: 10.250.112.0/22
+  enableVPCFlowLogs: false
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/values.yaml]: This is not needed

```suggestion
```

File: controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf [hunk 3]
```diff
@@ -29,6 +29,13 @@ resource "google_compute_subnetwork" "subnetwork-nodes" {
   ip_cidr_range = "{{ required "networks.workers is required" .Values.networks.workers }}"
   network       = "{{ required "vpc.name is required" .Values.vpc.name }}"
   region        = "{{ required "google.region is required" .Values.google.region }}"
+{{- if .Values.networks.enableVPCFlowLogs }}
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf]:
```suggestion
{{- if .Values.networks.flowLogs }}
```

(same below)

File: controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go [hunk 4]
```diff
@@ -42,6 +42,8 @@ type NetworkConfig struct {
 	Worker string
 	// Workers is the worker subnet range to create (used for the VMs).
 	Workers string
+	// FlowLog is the vpc flow log config to configure
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go]: Can we please find a better doc string?

File: controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go [hunk 5]
```diff
@@ -42,6 +42,8 @@ type NetworkConfig struct {
...
+	FlowLog *FlowLog
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go]:
```suggestion
	FlowLogs *FlowLogs
```

File: controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go [hunk 6]
```diff
@@ -104,3 +106,13 @@ type CloudNAT struct {
 	// The default value is 2048 ports.
 	MinPortsPerVM *int32
 }
+
+// FlowLog contains the configuration options for the vpc flow logs
+type FlowLog struct {
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/types_infrastructure.go]:
```suggestion
type FlowLogs struct {
```

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 7]
```diff
@@ -45,6 +45,8 @@ type NetworkConfig struct {
 	Worker string `json:"worker"`
 	// Workers is the worker subnet range to create (used for the VMs).
 	Workers string `json:"workers"`
+	// FlowLog is the vpc flow log config to configure
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]:
```suggestion
	// FlowLog is the vpc flow log config to configure
	// +optional
```

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 8]
```diff
@@ -45,6 +45,8 @@ type NetworkConfig struct {
...
+	FlowLog *FlowLog `json:"flowLog,omitempty"`
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]:
```suggestion
	FlowLogs *FlowLogs `json:"flowLogs,omitempty"`
```

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 9]
```diff
@@ -109,3 +111,13 @@ type CloudNAT struct {
 	// +optional
 	MinPortsPerVM *int32 `json:"minPortsPerVM,omitempty"`
 }
+
+// FlowLog contains the configuration options for the vpc flow logs
+type FlowLog struct {
+	// AggregationInterval for collecting flow logs
+	AggregationInterval string `json:"aggregationInterval,omitempty"`
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]: Is this an optional or a required field? If optional, then it must be a `*string` and `// +optional` tag should be there. If required, then remove `omitempty`.

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 10]
```diff
@@ -109,3 +111,13 @@ type CloudNAT struct {
...
+	// FlowSampling sets the sampling rate of VPC flow logs within the subnetwork where 1.0 means all collected logs are reported and 0.0 means no logs are reported
+	FlowSampling *float32 `json:"flowSampling,omitempty"`
+	// Metadata configures whether metadata fields should be added to the reported VPC flow logs
+	Metadata string `json:"metadata,omitempty"`
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]: Is this an optional or a required field? If optional, then it must be a `*string` and `// +optional` tag should be there. If required, then remove `omitempty`.

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 11]
```diff
@@ -109,3 +111,13 @@ type CloudNAT struct {
 	// +optional
 	MinPortsPerVM *int32 `json:"minPortsPerVM,omitempty"`
 }
+
+// FlowLog contains the configuration options for the vpc flow logs
+type FlowLog struct {
+	// AggregationInterval for collecting flow logs
+	AggregationInterval string `json:"aggregationInterval,omitempty"`
+	// FlowSampling sets the sampling rate of VPC flow logs within the subnetwork where 1.0 means all collected logs are reported and 0.0 means no logs are reported
+	FlowSampling *float32 `json:"flowSampling,omitempty"`
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]: Is this an optional or a required field? If optional, then `// +optional` tag should be there. If required, then remove `omitempty` and make it a `float32`.

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go [hunk 12]
```diff
@@ -117,10 +133,16 @@ func ComputeTerraformerChartValues(
...
 		"networks": map[string]interface{}{
-			"pods":     extensionscontroller.GetPodNetwork(cluster),
-			"services": extensionscontroller.GetServiceNetwork(cluster),
-			"workers":  workersCIDR,
-			"internal": config.Networks.Internal,
+			"pods":              extensionscontroller.GetPodNetwork(cluster),
+			"services":          extensionscontroller.GetServiceNetwork(cluster),
+			"workers":           workersCIDR,
+			"internal":          config.Networks.Internal,
+			"enableVPCFlowLogs": enableVPCFlowLogs,
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]:
```suggestion
```

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go [hunk 13]
```diff
@@ -117,10 +133,16 @@ func ComputeTerraformerChartValues(
...
+			"flowLog": map[string]interface{}{
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]:
```suggestion
			"flowLogs": map[string]interface{}{
```

Can you please only provide this to the Helm chart if the `InfrastructureConfig` contains config for the flow logs? If not, simply omit/don't pass it.

File: controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf [hunk 14]
```diff
@@ -29,6 +29,13 @@ resource "google_compute_subnetwork" "subnetwork-nodes" {
   ip_cidr_range = "{{ required "networks.workers is required" .Values.networks.workers }}"
   network       = "{{ required "vpc.name is required" .Values.vpc.name }}"
   region        = "{{ required "google.region is required" .Values.google.region }}"
+{{- if .Values.networks.flowLogs }}
+  log_config {
+    {{ if .Values.networks.flowLogs.aggregationInterval -}}aggregation_interval = "{{ .Values.networks.flowLogs.aggregationInterval }}"{{- end }}
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf]:
```suggestion
    {{ if .Values.networks.flowLogs.aggregationInterval }}aggregation_interval = "{{ .Values.networks.flowLogs.aggregationInterval }}"{{ end }}
```

File: controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf [hunk 15]
```diff
@@ -29,6 +29,13 @@ resource "google_compute_subnetwork" "subnetwork-nodes" {
...
+    {{ if .Values.networks.flowLogs.flowSampling -}}flow_sampling        = "{{ .Values.networks.flowLogs.flowSampling }}"{{- end }}
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf]:
```suggestion
    {{ if .Values.networks.flowLogs.flowSampling }}flow_sampling        = "{{ .Values.networks.flowLogs.flowSampling }}"{{ end }}
```

File: controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf [hunk 16]
```diff
@@ -29,6 +29,13 @@ resource "google_compute_subnetwork" "subnetwork-nodes" {
...
+    {{ if .Values.networks.flowLogs.metadata -}}metadata             = "{{ .Values.networks.flowLogs.metadata }}"{{- end }}
```
Comment (rfranzke) [inline controllers/provider-gcp/charts/internal/gcp-infra/templates/main.tf]:
```suggestion
    {{ if .Values.networks.flowLogs.metadata }}metadata             = "{{ .Values.networks.flowLogs.metadata }}"{{ end }}
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 17]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
 The specified CIDR ranges must be contained in the VPC CIDR specified above, or the VPC CIDR of your already existing VPC.
 You can freely choose these CIDRs and it is your responsibility to properly design the network layout to suit your needs.
 
+The `networks.flowLogs` section describes the configuration for the vpc flow logs. In order to enable the vpc flow logs at least one of the following parameters needs to be specified in the flow log section.
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
The `networks.flowLogs` section describes the configuration for the VPC flow logs. In order to enable the VPC flow logs, at least one of the following parameters needs to be specified in the flow log section:
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 18]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
...
+
+The `networks.flowLogs.aggregationInterval` is optional and describes the aggregation interval for collecting flow logs.
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
* `networks.flowLogs.aggregationInterval`, an optional parameter describing the aggregation interval for collecting flow logs.
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 19]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
...
+
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 20]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
...
+The `networks.flowLogs.flowSampling` is optional and sets the sampling rate of VPC flow logs within the subnetwork where 1.0 means all collected logs are reported and 0.0 means no logs are reported.
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
* `networks.flowLogs.flowSampling`, an optional parameter setting the sampling rate of VPC flow logs within the subnetwork where `1.0` means all collected logs are reported and `0.0` means no logs are reported.
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 21]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
...
+
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
```

File: controllers/provider-gcp/docs/usage-as-end-user.md [hunk 22]
```diff
@@ -62,6 +66,14 @@ The `networks.cloudNAT.minPortsPerVM` is optional and is used to define the [min
...
+The `networks.flowLogs.metadata` is optional and describes whether metadata fields should be added to the reported VPC flow logs.
```
Comment (rfranzke) [inline controllers/provider-gcp/docs/usage-as-end-user.md]:
```suggestion
* `networks.flowLogs.metadata`, an optional parameter describing whether metadata fields should be added to the reported VPC flow logs.
```

File: controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go [hunk 23]
```diff
@@ -45,6 +45,8 @@ type NetworkConfig struct {
 	Worker string `json:"worker"`
 	// Workers is the worker subnet range to create (used for the VMs).
 	Workers string `json:"workers"`
+	// FlowLogs contains the flow log configuration for the subnet.
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/apis/gcp/v1alpha1/types_infrastructure.go]:
```suggestion
	// FlowLogs contains the flow log configuration for the subnet.
	// +optional
```

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go [hunk 24]
```diff
@@ -144,8 +144,31 @@ func RenderTerraformerChart(
...
 ) (*TerraformFiles, error) {
+	var (
+		aggregationInterval string
+		flowSampling        float32
+		metadata            string
+	)
+
 	values := ComputeTerraformerChartValues(infra, account, config, cluster)
 
+	if config.Networks.FlowLogs != nil {
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: Can you also implement a unit test for this case, please?

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go [hunk 25]
```diff
@@ -144,8 +144,31 @@ func RenderTerraformerChart(
 	config *api.InfrastructureConfig,
 	cluster *extensionscontroller.Cluster,
 ) (*TerraformFiles, error) {
+	var (
+		aggregationInterval string
+		flowSampling        float32
+		metadata            string
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: Looking at the code below, I don't get the purpose of these variables. I don't think they are needed at all.

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go [hunk 26]
```diff
@@ -144,8 +144,31 @@ func RenderTerraformerChart(
...
+		}
+		if config.Networks.FlowLogs.Metadata != nil {
+			metadata = *config.Networks.FlowLogs.Metadata
+		}
+		values["networks"].(map[string]interface{})["flowLogs"] = map[string]interface{}{
+			"aggregationInterval": aggregationInterval,
+			"flowSampling":        flowSampling,
+			"metadata":            metadata,
+		}
+	}
```
Comment (rfranzke) [inline controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: Instead, I'd suggest something like this:

```suggestion
	if flowLogs := config.Networks.FlowLogs; flowLogs != nil {
	  var flowLogs map[string]interface{}
	  
	  if flowLogs.AggregationInterval != nil {
	    flowLogs["aggregationInterval"] = *flowLogs.AggregationInterval
	  }
	  
	  if flowLogs.FlowSampling != nil {
	    flowLogs["flowSampling"] = *flowLogs.FlowSampling
	  }
	  
	  if flowLogs.Metadata != nil {
	    flowLogs["metadata"] = *flowLogs.Metadata
	  }
	  
	  values["networks"].(map[string]interface{})["flowLogs"] = flowLogs
	}
```

**Comment by rfranzke** [comment]
Also, can you please enhance the release note so that it is more meaningful for end-users? #520 is a good example.

File: controllers/provider-gcp/pkg/internal/infrastructure/terraform.go (hunk 24 above)
Comment (rfranzke) [reply controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: This comment is not yet resolved, please check.

Comment (DockToFuture) [reply controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: Didn't we cover that with our unit tests when the flow log config was provided with wrong or correct values?

Comment (rfranzke) [reply controllers/provider-gcp/pkg/internal/infrastructure/terraform.go]: This is a different code path. What should be unit tested here is that the Helm chart values are correctly computed before they are handed over to Terraform.

**Comment by ialidzhikov** [comment]
What about 
> Requires v="3.4.0" of google terraform provide

I don't see a release of g/terraformer that includes this.
cc @rfranzke

**Comment by rfranzke** [comment]
https://github.com/gardener/gardener-extensions/pull/579#pullrequestreview-347802787

**Pull request merged**